### PR TITLE
[enhance](fs) Eliminate the reference of FileReader/FileWriter to FS

### PR DESCRIPTION
**Pull request by platoneko**
## Proposed changes

- Eliminate the reference of `FileReader/FileWriter` to FS
- Fix the bug where `BrokerFileSystem` does not set resource id
- Remove `create_empty_file` from the `FileWriterOptions`, and make it the default behavior

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by platoneko** [comment]
run buildall

File: be/src/io/fs/s3_file_writer.cpp [hunk 1]
```diff
@@ -207,7 +204,7 @@ Status S3FileWriter::close() {
             auto* buf = dynamic_cast<UploadFileBuffer*>(_pending_buf.get());
             DCHECK(buf != nullptr);
             buf->set_upload_to_remote([this](UploadFileBuffer& b) { _put_object(b); });
-        } else if (_create_empty_file) {
+        } else {
```
Comment (gavinchou) [inline be/src/io/fs/s3_file_writer.cpp:207]: buggy if _pending_buf == nullptr and it has been uploaded by once

Comment (gavinchou) [reply be/src/io/fs/s3_file_writer.cpp:207]: add a new flag to resolve this bug.

**Review by gavinchou — CHANGES_REQUESTED** [review]

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/inverted_index_compound_directory.cpp [hunk 2]
```diff
@@ -399,8 +399,7 @@ void DorisCompoundDirectory::FSIndexInput::readInternal(uint8_t* b, const int32_
 
 void DorisCompoundDirectory::FSIndexOutput::init(const io::FileSystemSPtr& fileSystem,
                                                  const char* path) {
-    io::FileWriterOptions opts {.create_empty_file = false};
-    Status status = fileSystem->create_file(path, &_writer, &opts);
+    Status status = fileSystem->create_file(path, &_writer);
```
Comment (morningman) [inline be/src/olap/rowset/segment_v2/inverted_index_compound_directory.cpp:402]:
```suggestion
    Status status = fs->create_file(path, &_writer);
```

File: be/src/olap/rowset/segment_v2/inverted_index_compound_directory.cpp [hunk 3]
```diff
@@ -512,11 +511,7 @@ void DorisCompoundDirectory::FSIndexOutput::close() {
 
 int64_t DorisCompoundDirectory::FSIndexOutput::length() const {
     CND_PRECONDITION(_writer != nullptr, "file is not open");
-    int64_t ret;
-    if (!_writer->fs()->file_size(_writer->path(), &ret).ok()) {
-        return -1;
-    }
-    return ret;
+    return _writer->bytes_appended();
```
Comment (morningman) [inline be/src/olap/rowset/segment_v2/inverted_index_compound_directory.cpp:514]: The default value of `_bytes_appended` is 0,
but the origin logic may return -1, need to check if `-1` means something to the caller.

File: be/src/olap/rowset/vertical_beta_rowset_writer_helper.cpp [hunk 4]
```diff
@@ -197,9 +197,7 @@ Status VerticalBetaRowsetWriterHelper::_create_segment_writer(
         return Status::Error<INIT_FAILED>("get fs failed");
     }
     io::FileWriterPtr file_writer;
-    io::FileWriterOptions opts;
-    opts.create_empty_file = false;
-    Status st = fs->create_file(path, &file_writer, &opts);
+    Status st = fs->create_file(path, &file_writer);
```
Comment (morningman) [inline be/src/olap/rowset/vertical_beta_rowset_writer_helper.cpp:200]: the default behavior of the file writer will create empty file.
But here we should not create empty file, so why remove `opts.create_empty_file = false;`?

File: be/src/runtime/snapshot_loader.cpp [hunk 5]
```diff
@@ -102,19 +102,16 @@ Status SnapshotLoader::init(TStorageBackendType::type type, const std::string& l
         S3URI s3_uri(location);
         RETURN_IF_ERROR(s3_uri.parse());
         RETURN_IF_ERROR(S3ClientFactory::convert_properties_to_s3_conf(_prop, s3_uri, &s3_conf));
-        std::shared_ptr<io::S3FileSystem> fs;
-        RETURN_IF_ERROR(io::S3FileSystem::create(std::move(s3_conf), "", &fs));
-        _remote_fs = std::move(fs);
+        _remote_fs =
```
Comment (morningman) [inline be/src/runtime/snapshot_loader.cpp:105]: Better to create a FsFactory class?
Maybe in next PR

File: be/src/olap/rowset/segment_v2/inverted_index_compound_directory.cpp (hunk 3 above)
Comment (platoneko) [reply be/src/olap/rowset/segment_v2/inverted_index_compound_directory.cpp:514]: Seems that there is no code within clucene that checks if the return value of `length` is -1.

**Review by gavinchou — APPROVED** [review]

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**